File: _utils/stack.py
import json


class Stack:
    def __init__(self) -> None:
        self.vals = []


    def push(self, val):
        """push the given value into the stack.
        Args:
            val     the value to push
        """
        self.vals.append(val)


    def pop(self):
        """pop the top value from the stack.
        Returns:
            the top value from the stack, or `None` if not available.
        """
        if len(self.vals) == 0:
            return None
        else:
            return self.vals.pop()


    def peek(self):
        """peek the top value from the stack.
        Returns:
            the top value from the stack, or `None` if not available.
        """
        if len(self.vals) == 0:
            return None
        else:
            return self.vals[-1]


    def empty(self):
        """Check whether the stack is empty or not.
        Returns:
            `True` if the stack is empty; otherwise `False`.
        """
        return len(self.vals) == 0


    def clear(self):
        """Clear the entire stack.
        """
        self.vals = []


    def __str__(self) -> str:
        """Cast the stack into a string representation.
        Returns:
            the string representation of the stack.
        """
        return f"{json.dumps(self.vals)}<="


def main() -> None:
    stack = Stack()
    print(f"{stack}; empty? {stack.empty()}")

    stack.push(1)
    print(f"{stack}; empty? {stack.empty()}")
    stack.push(3);
    print(f"{stack}; empty? {stack.empty()}")
    print(f"popperd: {stack.pop()}")
    print(f"{stack}; empty? {stack.empty()}")
    stack.push(5);
    print(f"{stack}; empty? {stack.empty()}")
    print(f"popperd: {stack.pop()}")
    print(f"{stack}; empty? {stack.empty()}")
    print(f"popperd: {stack.pop()}")
    print(f"{stack}; empty? {stack.empty()}")
    print(f"popperd: {stack.pop()}")
    print(f"{stack}; empty? {stack.empty()}")


# if __name__ == "__main__":
#     main()

File: _utils/stack.js
/**
 * The stack definition
 */
const Stack = module.exports.Stack = class {
    constructor() {
        this._vals = [];
    }


    /**
     * push the given value to the stack
     * @param {object} val  the value to push
     */
    push(val) {
        this._vals.push(val);
    }


    /**
     * pop the top value from the stack
     * @returns     the top value from the stack, or `null` if not available
     */
    pop() {
        return this._vals.length === 0 ? null : this._vals.pop();
    }


    /**
     * query the top value from the stack
     * @returns      the top value from the stack, or `null` if not available
     */
    peek() {
        return this._vals.length === 0 ? null : this._vals[this._vals.length - 1];
    }


    /**
     * check whether the stack is empty or not
     * @returns     `true` if stack is empty; otherwise `false`
     */
    empty() {
        return this._vals.length === 0;
    }


    /**
     * clear the entire stack to make it empty
     */
    clear() {
        this._vals = [];
    }


    /**
     * cast the stack into a string representation
     * @returns     the string representation of the stack
     */
    toString() {
        return `[${this._vals.join(", ")}]<=`;
    }
};


function main() {
    const stack = new Stack();
    console.log(`${stack}; empty? ${stack.empty()}`);

    stack.push(1);
    console.log(`${stack}; empty? ${stack.empty()}`);
    stack.push(3);
    console.log(`${stack}; empty? ${stack.empty()}`);
    console.log(`popped: ${stack.pop()}`);
    console.log(`${stack}; empty? ${stack.empty()}`);
    stack.push(5);
    console.log(`${stack}; empty? ${stack.empty()}`);
    console.log(`popped: ${stack.pop()}`);
    console.log(`${stack}; empty? ${stack.empty()}`);
    console.log(`popped: ${stack.pop()}`);
    console.log(`${stack}; empty? ${stack.empty()}`);
    console.log(`popped: ${stack.pop()}`);
    console.log(`${stack}; empty? ${stack.empty()}`);
}

// main();
